### PR TITLE
Improve a way to handle errors for collection nested attributes

### PR DESCRIPTION
**Pull request by bogdan**
Example:

``` ruby
    Employment.validates :company, :presence => true
    Developer.has_many :employments
    Developer.accepts_nested_attributes_for :employments
    d = Developer.new
    2.times { d.employments.build }
    d.valid?
    d.errors.messages
```

Old result:

``` ruby
    {
      :"employments.company"=>["can't be blank"],
      :email=>["can't be blank"],
      :password=>["can't be blank"]
    }

```

See that it is impossible to detect which `Employment` object is error related to.
In order to fix that propose to use an index(as Employment could be new - we can not use an id).
Only applied to collection associations.

New result

``` ruby
    {
      :"employments[0].company"=>["can't be blank"],
      :"employments[1].company"=>["can't be blank"],
      :email=>["can't be blank"],
      :password=>["can't be blank"]
    }

```

Live demo:
http://rails-ajax-validation.herokuapp.com

### CODE REVIEW

**Comment by dmitry** [comment]
Anyone is going to work on it?

I agree that the errors should follow an object convention, not flat hashmap that should be additionally parsed afterwards.

I've implemented 3 years ago almost the same functionality in Rails 3.2: https://gist.github.com/dmitry/3238745 (ugly, but works perfectly)

I would like to participate in development of this feature if it's possible.

**Comment by bogdan** [comment]
Sum up discussion above:

There is currently open question regarding migration plan, here are the options:

1. No backward compatibility - yew
2. Ability to have a single configuration option to change the behaviour for entire project
3. Configure per `accepts_nested_attributes_for` call with `index_errors: true/false` option (with further deprecation or may be support forever)

Also there is an open discussion regarding errors format:

1. Use flat key/value errors format like in current PR
2. Use nested hashes structure

``` ruby
{"educations[1].company" =≥ "can not be blank"} #1
{ "educations" => [{}, {:company => "can not be blank"}]} #2
```

My personal choice is:
compatibility - 3 - this is the most smooth way allowing to migrate usages one by one
format - 1 - it is more backward compatible and `json <-> keyvalue` conversion algorithm is already implemented on many platforms including Rails and jQuery. 

@dmitry you can help by putting us your arguments on choices.

**Comment by tsun1215** [comment]
@bogdan I've actually got something working for this, but I've been trying to figure out creating models and test cases for it (new to contributing). Here's what I have (work in progress): https://github.com/tsun1215/rails/commit/68bb7e7a589247ad97d64318d1c51d710e1e6123; With the manual testing I did, it should be working. (Also, the puts statement will be removed when I actually submit a request).

**Comment by bogdan** [comment]
@tsun1215 yes I saw your work, that is why I've mansion this variant here. Thanks for showing me it.

**Comment by sobrinho** [comment]
In our side we ended making a errors presenter:

``` ruby
class NestedErrorsPresenter
  attr_reader :object, :klass

  def initialize(object)
    @object = object
    @klass = object.class
  end

  def to_h
    errors = HashWithIndifferentAccess.new

    klass.reflect_on_all_associations.each do |association|
      if association.macro == :has_many && association.options[:autosave]
        object.send(association.name).each_with_index do |child, index|
          next if child.valid?

          index = child.respond_to?(:front_end_index) ? child.front_end_index : index

          errors[association.name] = {}
          errors[association.name][:records_errors] ||= {}
          errors[association.name][:records_errors][index] = child.errors.to_h
        end
      end
    end

    object.errors.each do |attribute, error|
      next if attribute =~ /\./

      if klass.reflect_on_association(attribute)
        if association = klass.reflect_on_association(attribute)
          errors[attribute] ||= {}
          errors[attribute][:relation_errors] = error

          if association.macro == :has_one
            errors[attribute][:record_errors] = object.send(attribute).errors.to_h
          end
        else
          errors[attribute] = error
        end
      else
        errors[attribute] = error
      end
    end

    { :errors => errors }
  end
end
```

And using a custom reponder:

``` ruby
class NestedErrorResponder < ApplicationResponder
  def to_json
    if resource.valid?
      render :json => resource
    else
      render :json => NestedErrorsPresenter.new(resource).to_h, :status => :unprocessable_entity
    end
  end
end
```

It's working perfectly fine on production but we are still expecting this feature to be on core since it's a common need.

If not, we are going to pack this as gem and publish :)

**Comment by dmitry** [comment]
There are actually 3 types of error formats possible:

- flat: `{"educations[1].company" =≥ "can not be blank"}`
- nested with array index: `{"educations": [{}, {"company" =≥ "can not be blank"}]}`
- nested with hash index: `{"educations": {"1": {"company" =≥ "can not be blank"}}}`

I prefer nested with hash index (3). It's more universal and can be used in any languages, sometimes can be faster.

*Flat*

- can be faster in some languages/environemnts (like javascript)
- must be parsed before it can be applied to (not all the languages have the parsing feature, they should be implemented)

*Nested with array index*

- not possible to have index with high value (sometimes it's may be required for the values objects)
- can be parsed in any languages that have hashmap and array types

*Nested with hash index*

- pros from *nested with array index*
- only hash/object is used
- possible to have any index, including string

**Comment by sobrinho** [comment]
Please, be aware that we must support this structure:

``` javascript
{
  "name": "can't be blank",
  "partners": {
    "record_errors": {
      0: { "percent": "can't be greater than 100" },
      3: { "percent": "can't be greater than 100" }
    },
    "relation_errors": {
      "base": "the total of percentages can't be greater than 100"
    }
  }
}
```

First the index is not always sequential that is there is an error on first record and third record, so, it must be always a hash instead of an array.

Second we must provide the subitems `record_errors` which contains the nested record errors and `relation_errors` which are the errors on the relation itself.

And honestly I think that the plain hash will be always worse since we will need to parse the keys everywhere and in every language (we don't always use this return on browser, it may be a mobile app or another program in Go/Python/Scala/etc) and returning nested hashes will avoid that.

**Comment by dmitry** [comment]
@sobrinho :+1: with small exception: subitems, because we can follow convention to use indexes of the records as numbers, and general errors with index as strings (normally it's just `base` index string). Plus base can be an array of strings.

**Comment by alexspeller** [comment]
Surely all errors can be arrays of strings, as there can be multiple errors on any field

**Comment by sobrinho** [comment]
I sent a wrong hash, sorry. The correct one would be:

``` javascript
{
  "name": ["can't be blank"],
  "partners": {
    "record_errors": {
      0: { "percent": ["can't be greater than 100"] },
      3: { "percent": ["can't be greater than 100"] }
    },
    "relation_errors": ["the total of percentages can't be greater than 100"]
  }
}
```

Note that in a case of a has one association, the hash would be different:

``` javascript
{
  "name": ["can't be blank"],
  "address": {
    "record_errors": {
      "street": ["can't be blank"]
    },
    "relation_errors": ["is outside the country"]
  }
}
```

But still we must have `record_errors` and `relation_errors` to represent the association/children association errors and the errors on the association itself.

Note that I'm also proposing the messages to be always an array now, it caused some trouble to us to have sometimes a string and sometimes an array, keeping it always as array make the front-end logic simpler.

**Comment by bogdan** [comment]
I am not sure what kind of architecture changes will be under all these relation errors.
Current code is hardly tided on a fact that errors model structure is flat. The `Errors#add` method itself is not designed to accept hash argument.

If the nested errors should be supported, it should be done at a different layer for example custom `ARBase` method like `#errors_with_associations_as_json` (better name welcome).

Allow `AM::Errors` to act like a hash with possibly infinite nesting is very very radical change.

**Comment by sobrinho** [comment]
`Errors#add` wouldn't change, I'm generating this proposed hash with the current hash using this presenter:

``` ruby
class NestedErrorsPresenter
  attr_reader :object, :klass

  def initialize(object)
    @object = object
    @klass = object.class
  end

  def to_h
    errors = HashWithIndifferentAccess.new

    klass.reflect_on_all_associations.each do |association|
      if association.macro == :has_many && association.options[:autosave]
        object.send(association.name).each_with_index do |child, index|
          next if child.valid?

          index = child.respond_to?(:front_end_index) ? child.front_end_index : index

          errors[association.name] = {}
          errors[association.name][:records_errors] ||= {}
          errors[association.name][:records_errors][index] = child.errors.to_h
        end
      end
    end

    object.errors.each do |attribute, error|
      next if attribute =~ /\./

      if klass.reflect_on_association(attribute)
        if association = klass.reflect_on_association(attribute)
          errors[attribute] ||= {}
          errors[attribute][:relation_errors] = error

          if association.macro == :has_one
            errors[attribute][:record_errors] = object.send(attribute).errors.to_h
          end
        else
          errors[attribute] = error
        end
      else
        errors[attribute] = error
      end
    end

    { :errors => errors }
  end
end
```

So, the Errors object wouldn't change, only the presentation of it.

**Comment by bogdan** [comment]
@sobrinho probably you didn't but others(including myself) made me think that this is proposed.
I agree with you that it should be decoupled from `AM::Errors`.

I don't feel like it is related with this PR as it only cares about how errors are added to `Errors` object.
as I said before: Rails is not doing it the right way.

Presenters like yours can live beside this PR in Rails Core or outside of it.

**Comment by sobrinho** [comment]
@bogdan not sure if I'm following you but the suggestion above is one solution for the problem of this pull request.

People seems to agree, everyone on this pull request at least, that we must have a way to deal with it on active record natively.

A presenter may or may not be the best solution but thinking in backwards compatibility it seems to be the way.

Are we saying the same thing or I did not understood you?

My suggestion is:

1. Keep `ActiveModel::Errors` as it is today to not break anything
2. Implement a `ActiveModel::ErrorsPresenter` which computes a hash like I suggested or like you suggested first

--

I'm suspect to say that but the hash which I proposed is working perfectly on at least 5 large applications here.

Our front-end architecture is made by 4 javascript packages (private but will be open sourced in a couple of weeks):

**backbone.relation**: Handles the concept of associations in backbone models (similar to Backbone.Relational)

**backbone.errors**: Handle the errors hash on failure and distributes the errors through associations (using a small reflection implemented on backbone.relation)

**nested-errors-presenter**: Handle the errors receiving the attribute name and the error and render it on DOM (assuming that you use bootstrap markup but it accepts any strategy)

**backbone.nested-errors-presenter**: Basically knows how to receive the errors from a backbone.relation model and iterate through them and send it to nested-errors-presenter.

Basically using these 4 packages the back-ends validations are totally painless to integrate on backbone and render them to DOM.

We started working with React.JS right now and we are going to create React components/libraries which will knows how to deal with this hash too.
